### DISCUSSION TITLE
FUNCTION_INVOCATION_TIMEOUT in getServerSideProps - MongoDB

### DISCUSSION THREAD

**Discussion by ItsJohnTheDeveloper** [prisma/prisma#15948] [Q&A]
Hello Prisma devs & community :wave: 

I'm getting a `FUNCTION_INVOCATION_TIMEOUT` error in the serverless function deployed on Vercel. 
Long story short I'm fetching data using the prisma client inside `getServerSideProps` . Its taking longer than 10 seconds to instantiate a db connection, and query data. Could it have something to do with the connection to MongoDB not being cached? 

_*please note: using the prisma client in getServerSideProps works great in a LOCAL environment, but NOT in a SERVERLESS environment - as it will timeout._

**code block** 
```jsx
import prisma from "../prisma";

export const getServerSideProps = async ({ params }) => {
  const artist: ArtistDocument = await prisma.user.findUnique({
    where: { id: params.artist_id },
  });
  return {
    props: {
      data: {
        artist
      },
    },
  };
};
```

```js
// prisma.ts
import { PrismaClient } from "@prisma/client";

// add prisma to the NodeJS global type
type CustomGlobal = Global & {
  prisma: PrismaClient;
};

// Prevent multiple instances of Prisma Client in development
declare const global: CustomGlobal;
const prisma =
  global.prisma ||
  new PrismaClient({ log: ["query", "info", "warn", "error"] });

// re-use existing prisma client on consecutive hot reloads
if (process.env.NODE_ENV === "development") {
  global.prisma = prisma;
  console.log("Using cached connection.");
}

export default prisma;
```

A temporary workaround solution I found was to just query the data client side, hitting a nextjs api endpoint that queries data from mongodb via prisma client.

### Environment & setup
node 16
- package.json
"@next-auth/prisma-adapter": "^1.0.3",
"@prisma/client": "^3.13.0",
"prisma": "^3.13.0",
"mongodb": "^4.6.0",
"next": "12.0.10",
"react": "17.0.2",

### Error
![Screenshot from 2022-10-22 18-36-39](https://user-images.githubusercontent.com/40045461/197417565-c9da55d0-2677-486f-a573-eb02bf39b19f.png)

### Vercel log error
![Screenshot from 2022-10-22 18-40-00](https://user-images.githubusercontent.com/40045461/197417560-fda277db-ee24-4444-8fc3-50cf7e520ce6.png)

If I'm missing any details that will help further debug this issue, please let me know and i'll do my best to provide more detail.
****

**Comment by austincrim** [comment]
Hey @ItsJohnTheDeveloper 👋,

Do you know how long the same query takes when running locally?

Could you enable [query logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#log-to-stdout) and provide the output here?

**Reply by ItsJohnTheDeveloper** [reply]
Hi @austincrim :wave: Thanks for your response.

I noticed while adding the logs, the docs mentioned: 
>Note that for MongoDB, the params and duration fields will be undefined.

**Reply by ItsJohnTheDeveloper** [reply]
Do you have any ideas how I'd be able to further debug this?

**Reply by austincrim** [reply]
You could use `console.time()` and `console.timeEnd()` to get an estimate of how long the query takes locally. I'm not sure why the difference in duration would be so great between local and deployed environments.

**Reply by ItsJohnTheDeveloper** [reply]
Could this be an issue with MongoDB's connection pooling? I'll provide more code samples of my project to help debug I'm just not sure what else to provide at this point. 

Conn string: 
`mongodb+srv://<user>:<pass>@<cluster>.pixym.mongodb.net/<db>?maxPoolSize=1`

**Reply by austincrim** [reply]
If you could provide a minimal reproduction repo, I can dig in further 👍

**Reply by ItsJohnTheDeveloper** [reply]
Yeah for sure! 
https://github.com/ItsJohnTheDeveloper/artshow.me

**Reply by ItsJohnTheDeveloper** [reply]
@austincrim Any luck?

**Reply by pantharshit00** [reply]
Hey @ItsJohnTheDeveloper 

Just to make sure, can you check in region your database is hosted. From the above information, looks like your serverless function is invoking in Portland so make sure your database is also close by.

**Reply by ItsJohnTheDeveloper** [reply]
Hey @pantharshit00 
Its set to west coast right now: `AWS / Oregon (us-west-2)`

**Reply by pantharshit00** [reply]
Hmm, that is quite weird then. One other thing that I will request is can you also confirm this happens with the latest version of Prisma?

**Reply by ItsJohnTheDeveloper** [reply]
I can confirm this is happening on the latest version of Prisma and MongoDB.

**Reply by pantharshit00** [reply]
Okay! I looked at your application and everything looks fine at a first look. You can maybe make some queries in parallel using `Promise.all` but otherwise, it looks fine.

Can you please tell me which route is experiencing this timeout or this is happening with every route? Right now I wasn't able to pin point that out. 

Also, can you check the connections graph in metrics of MongoDB atlas and see it's state.

**Reply by ItsJohnTheDeveloper** [reply]
This is happening on every route. I even made a test route and simply just importing and declarding prismaclient inside _getServerSideProps_ caused it to timeout :thinking: 

### graph metrics: 
It spun up 6 connections for some reason. I visited a couple different deployment urls and refreshing the page a few times. 
![Screenshot from 2022-12-01 10-10-01](https://user-images.githubusercontent.com/40045461/205128693-7c47345e-326a-4e2d-9c60-e81586ad1c45.png)

**Reply by nurul3101** [reply]
Can you try updating to the latest Prisma version and see if you still get these timeout errors?

**Reply by ItsJohnTheDeveloper** [reply]
@nurul3101 Can confirm this is happening on prisma version 4.7.1

**Reply by pantharshit00** [reply]
Hello @ItsJohnTheDeveloper,

Sorry for late follow up here, now I will recommend measuring which section of you application is slow here. Can you please try adding few [`console.time()`](https://developer.mozilla.org/en-US/docs/Web/API/console/time) statements in the code to see if instantiating the client is taking the most time or the query is slow. 

You can also try using the tracing feature which Prisma provides. https://www.prisma.io/docs/concepts/components/prisma-client/opentelemetry-tracing

**Reply by ItsJohnTheDeveloper** [reply]
Hello @pantharshit00, thanks for your reply. 

Any page in Nextjs that is importing the Prisma Client will immediately timeout. I could have a completely empty `getServerSideProps` with no queries at all and it will still timeout. 

Maybe Server-side data fetching in a serverless environment using Nextjs & MongoDB just isn't supported yet in Prisma :man_shrugging:

*Keep in mind I am using the _free_ tier in Vercel which has a limit of **10 seconds** with _serverless function execution time._ _Maybe_ this problem will just **solve itself** if I pay for Vercel and have **more** execution time

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
